pyccel/parser/semantic.py

```diff
@@ -2511,14 +2519,13 @@ def _visit_FunctionDef(self, expr, **settings):
             func = FunctionDef(name, args, results, [])
             self.insert_function(func)
 
+            self._allocs.append([])
             # we annotate the body
             body = self._visit(expr.body)
 
-            #checking if the Function don't have Return then call the garbage collector
-            if not isinstance(expr.body.args[-1][-1], Return):
-                dealloc = [Deallocate(i) for i in self._allocs]
-                body = CodeBlock([body, CodeBlock(dealloc)])
-            self._allocs = []
+            #Call to the garbage collector
+            body = self.garbage_collector(body)
+            self._allocs = self._allocs[:-1]
```

Comment:
See : https://www.techiedelight.com/remove-last-element-from-list-python/
for a discussion about the best way to remove one element form the end of a list